### PR TITLE
Add new variables to set subnet by name

### PR DESCRIPTION
**Pull request by jiaweitao001**
## Issue number

To resolve issue #36 https://github.com/Azure/terraform-azurerm-loadbalancer/pull/36

## Checklist before requesting a review
- [ ] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [ ] I have executed pre-commit on my machine
- [ ] I have passed pr-check on my machine

Thanks for your cooperation!

### CODE REVIEW

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @jiaweitao001 for opening this pr! There're some issues need to be solved before we could merge this commit.

File: locals.tf [hunk 1]
```diff
@@ -1,4 +1,5 @@
 locals {
-  lb_name  = var.name != "" ? var.name : format("%s-lb", var.prefix)
-  pip_name = var.pip_name != "" ? var.pip_name : format("%s-publicIP", var.prefix)
+  lb_name         = var.name != "" ? var.name : format("%s-lb", var.prefix)
+  pip_name        = var.pip_name != "" ? var.pip_name : format("%s-publicIP", var.prefix)
+  vnet_id_by_name = length(data.azurerm_subnet.snet) == 1 ? element(data.azurerm_subnet.snet[*].id, 0) : ""
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply locals.tf:4]: Will modify.

File: main.tf [hunk 2]
```diff
@@ -38,7 +46,7 @@ resource "azurerm_lb" "azlb" {
     private_ip_address            = var.frontend_private_ip_address
     private_ip_address_allocation = var.frontend_private_ip_address_allocation
     public_ip_address_id          = try(azurerm_public_ip.azlb[0].id, "")
-    subnet_id                     = var.frontend_subnet_id
+    subnet_id                     = var.frontend_subnet_id != "" ? var.frontend_subnet_id : local.vnet_id_by_name
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply main.tf:49]: It should be `subnet_id_by_name`.

[the remark this answers is not in the document]
Comment (jiaweitao001) [reply main.tf:49]: Right, I think maybe we even don't need this expression in `locals.tf`, we can directly put `subnet_id = coalesce(var.frontend_subnet_id, try(data.azurerm_subnet.snet[0].id, ""))` in `main.tf`.

File: test/e2e/terraform_e2e_test.go [hunk 3]
```diff
@@ -7,8 +7,17 @@ import (
 	"github.com/gruntwork-io/terratest/modules/terraform"
 )
 
-func TestExamplesStartup(t *testing.T) {
-	test_helper.RunE2ETest(t, "../../", "examples/startup", terraform.Options{
-		Upgrade: true,
-	}, func(t *testing.T, output test_helper.TerraformOutput) {})
+func TestExmaples(t *testing.T) {
+	testPaths := []string{
+		//"examples/startup",
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply test/e2e/terraform_e2e_test.go:12]: My mistake, forgot to remove this before commit. Will do.

File: test/e2e/terraform_e2e_test.go [hunk 4]
```diff
@@ -7,8 +7,17 @@ import (
...
+		"examples/setsubnetbyname",
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply test/e2e/terraform_e2e_test.go:13]: Sure.

File: variables.tf [hunk 5]
```diff
@@ -39,6 +39,18 @@ variable "frontend_subnet_id" {
   default     = ""
 }
 
+variable "frontend_subnet_name" {
+  description = "(Optional) Frontend virtual network name to use when in private mode. Ignored if `frontend_subnet_id` is set."
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply variables.tf:43]: Sure. Will do.

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @jiaweitao001 for updating this pr, we still have some issues to solve.

File: test/unit/terraform_loadbalancer_test.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"testing"
+
+	test_helper "github.com/Azure/terraform-module-test-helper"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_FrontendSubnetIdNotExist(t *testing.T) {
+	vars := map[string]interface{}{
```
Comment (lonegunmanb) [inline test/unit/terraform_loadbalancer_test.go:12]: Why do we set `frontend_subnet_id` in `Test_FrontendSubnetIdNotExist` test method?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @jiaweitao001 for the update, we still have some issues to solve.

File: main.tf [hunk 7]
```diff
@@ -38,7 +50,14 @@ resource "azurerm_lb" "azlb" {
     private_ip_address            = var.frontend_private_ip_address
     private_ip_address_allocation = var.frontend_private_ip_address_allocation
     public_ip_address_id          = try(azurerm_public_ip.azlb[0].id, "")
-    subnet_id                     = var.frontend_subnet_id
+    subnet_id                     = local.subnet_id
+  }
+
+  lifecycle {
+    precondition {
+      condition     = var.frontend_subnet_name != "" || var.frontend_subnet_id != ""
```
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply main.tf:58]: My bad. It supposed to be `var.frontend_subnet_name == "" || var.frontend_subnet_id == ""`.

File: variables.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (jiaweitao001) [reply variables.tf:43]: Sure. Will change the description.

File: test/unit/terraform_loadbalancer_test.go (hunk 6 above)
Comment (jiaweitao001) [reply test/unit/terraform_loadbalancer_test.go:12]: It's testing against subnet id by name does not exist. Will change the test name.

File: main.tf (hunk 7 above)
[the remark this answers is not in the document]
Comment (lonegunmanb) [reply main.tf:58]: What if `var.frontend_subnet_name` is a valid subnet name, meanwhile `frontend_subnet_id` is `null`?

File: test/unit/terraform_loadbalancer_test.go [hunk 8]
```diff
@@ -8,7 +8,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func Test_FrontendSubnetIdNotExist(t *testing.T) {
+func Test_SubnetIdByNameNotExist(t *testing.T) {
```
Comment (lonegunmanb) [inline test/unit/terraform_loadbalancer_test.go:11]: I would recommend `Test_SubnetIdWithoutSubnetName`

**Review by lonegunmanb — CHANGES_REQUESTED** [review]

**Review by lonegunmanb — CHANGES_REQUESTED** [review]

**Review by lonegunmanb — APPROVED** [review]
Thanks @jiaweitao001, LGTM! 🚀

**Pull request merged**